src/azul/service/pfb.py

```diff
@@ -0,0 +1,445 @@
+
+    def entities(self, schema: JSON) -> Iterable[PFBEntity]:
+        for entity, relations in self._entities_to_relations.items():
+            normalize(entity, schema)
+            replace_null_with_empty_string(entity.object)
+            yield entity.to_json(relations)
+
+
+def normalize(entity: PFBEntity, schema):
+    """
+    Compare entities against the schema and add any fields that are missing.
+
+    None is the default value, but because of https://github.com/DataBiosphere/azul/issues/2370
+    this isn't currently reflected in the schema.
+    """
+    object_schema = one(f for f in schema['fields'] if f['name'] == 'object')
+    entity_schema = one(e for e in object_schema['type'] if e['name'] == entity.name)
+    fields = entity_schema['fields']
+    for field in fields:
+        if field['name'] not in entity.object:
+            if entity.name == 'files':
+                default_value = None
+            else:
+                assert field['type']['type'] == 'array'
+                # Default for a list of records is an empty list
+                if isinstance(field['type']['items'], dict):
+                    assert field['type']['items']['type'] == 'record'
+                    default_value = []
+                # All other types are lists of primitives where None is an accepted value
+                else:
+                    # Change 'string' to 'null' in https://github.com/DataBiosphere/azul/issues/2462
+                    assert 'string' in field['type']['items']
+                    default_value = [None]
+            entity.object[field['name']] = default_value
+
+
+def replace_null_with_empty_string(object_json: MutableJSON):
+    # FIXME remove with https://github.com/DataBiosphere/azul/issues/2462
+    for k, v in object_json.items():
+        if isinstance(v, dict):
+            replace_null_with_empty_string(v)
+        elif v is None:
+            object_json[k] = ''
+        elif v == [None]:
+            object_json[k] = ['']
+
+
+def write_entities(entities: Iterable[JSON], json_schema: JSON, path: str):
+    # fastavro doesn't know about our JSON type, hence the cast
+    parsed_schema = fastavro.parse_schema(cast(dict, json_schema))
+    with open(path, 'w+b') as fh:
+        # Writing the entities one at a time is ~2.5 slower, but makes it clear
+        # which entities fail, which is useful for debugging.
+        if config.debug:
+            log.info('Writing PFB entities individually')
+            for entity in entities:
+                try:
+                    fastavro.writer(fh, parsed_schema, [entity], validator=True)
+                except ValidationError:
+                    log.error('Failed to write Avro entity: %r', entity)
+                    raise
+        else:
+            fastavro.writer(fh, parsed_schema, entities, validator=True)
+
+
+def pfb_schema(encoded_schema: Iterable[JSON]) -> JSON:
```

Comment:
And am I correct in interpreting this as a skeleton schema that contains the pre-requisites for `encoded_schema`. Did you make this up or is this copied from somewhere? If so the source needs to be specified. I think this needs a docstring. I think it might also be useful to distinguish between values that are defined by Avro with values that are defined by PFP. We should discuss in parking lot.